### PR TITLE
virt: New virt_clone module for libvirt cloning capabilities

### PR DESCRIPTION
**Pull request by mrlesmithjr**
##### SUMMARY

This module is for a new virt_clone module for cloning libvirt VMs. By
using this module we can easily define a template(src) VM to create a
new VM by cloning. This module is idempotent and is check_mode aware.

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

virt_clone

##### ANSIBLE VERSION

```
ansible 2.5.1
  config file = None
  configured module search path = ['/Users/larry/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.6/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.6.5 (default, Mar 30 2018, 06:42:10) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]
```

##### ADDITIONAL INFORMATION

Example playbook:
```yaml
---
- hosts:          kvm
  vars:
    vm_template:  ubuntu1604-packer-template
    vms:
      - name:     app01
        state:    started
        template: "{{ vm_template }}"
      - name:     db01
        state:    started
        template: "{{ vm_template }}"
      - name:     lb01
        state:    started
        template: "{{ vm_template }}"
      - name:     web01
        state:    started
        template: "{{ vm_template }}"
  tasks:
    - name:       Cloning VMs
      virt_clone:
        name:     "{{ item['name'] }}"
        state:    "{{ item['state'] }}"
        template: "{{ item['template'] }}"
        uuid:     "{{ item['uuid']|default(omit) }}"
      become:     true
      loop:       "{{ vms }}"
```
Example execution:
```bash
PLAY [kvm] *********************************************************************************************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] *********************************************************************************************************************************************************************************************************************************************************************************************
ok: [nas02]

TASK [Cloning VMs] *************************************************************************************************************************************************************************************************************************************************************************************************
ok: [nas02] => (item={'name': 'app01', 'state': 'started', 'template': 'ubuntu1604-packer-template'})
changed: [nas02] => (item={'name': 'app02', 'state': 'started', 'template': 'ubuntu1604-packer-template'})
ok: [nas02] => (item={'name': 'db01', 'state': 'started', 'template': 'ubuntu1604-packer-template'})
ok: [nas02] => (item={'name': 'lb01', 'state': 'started', 'template': 'ubuntu1604-packer-template'})
ok: [nas02] => (item={'name': 'web01', 'state': 'started', 'template': 'ubuntu1604-packer-template'})

PLAY RECAP *********************************************************************************************************************************************************************************************************************************************************************************************************
nas02                      : ok=2    changed=1    unreachable=0    failed=0
```

### CODE REVIEW

**Comment by mrlesmithjr** [comment]
Anything else needed from me in regards to this? Finally got everything sorted out as I had some fun getting all of my rebases completed.

**Comment by bartekrutkowski** [comment]
@TimothyVandenbrande @UnderGreen @drybjed @flaper87 @helldorado @joshainglis @karmab @machacekondra @mwperina 
What's the status of this PR? Anything keeping it from being accepted?

**Comment by mrlesmithjr** [comment]
Have wondered the same ;)

File: lib/ansible/modules/cloud/misc/virt_clone.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Larry Smith Jr. <mrlesmithjr@gmail.com>
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/misc/virt_clone.py:4]:
```suggestion
# Copyright: (c) 2017, Larry Smith Jr. <mrlesmithjr@gmail.com>
```

File: lib/ansible/modules/cloud/misc/virt_clone.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+# Requirements:
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/misc/virt_clone.py]: Move this information to notes section, so user will be able to see using `ansible-doc` command. Just like [this](https://github.com/ansible/ansible/blob/0127a75132a6802526997496057235b893dcfcf1/lib/ansible/modules/cloud/vmware/vmware_guest.py#L31)

File: lib/ansible/modules/cloud/misc/virt_clone.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+author: Larry Smith Jr. (@mrlesmithjr)
+module: virt_clone
+short_description: Clones VMs from templates
+description:
+  - "This module creates clones for libvirt."
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/misc/virt_clone.py]:
```suggestion
  - This module creates clones for libvirt.
```

File: lib/ansible/modules/cloud/misc/virt_clone.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+version_added: "2.6"
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/misc/virt_clone.py]: s/2.6/2.8/

File: lib/ansible/modules/cloud/misc/virt_clone.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+options:
+  auto:
+    default: true
+    description:
+        - Generates a new guest name, and paths for new storage - default true
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/misc/virt_clone.py]:
```suggestion
        - Generates a new guest name, and paths for new storage - default true.
```

File: lib/ansible/modules/cloud/misc/virt_clone.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    type: bool
+  mac:
+    description:
+        - Allows a pre-defined MAC address to be assigned to the cloned VM
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/misc/virt_clone.py:39]:
```suggestion
        - Allows a pre-defined MAC address to be assigned to the cloned VM.
```

File: lib/ansible/modules/cloud/misc/virt_clone.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+  name:
+    aliases: [ "dest", "guest" ]
+    description:
+        - The VM to create
+    required: true
+  state:
+    choices: [ "present", "started" ]
+    default: present
+    description:
+        - If the clone should be present or started - default present
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/misc/virt_clone.py:49]:
```suggestion
        - If the clone should be present or started - default present.
```

File: lib/ansible/modules/cloud/misc/virt_clone.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+    required: false
+  sysprep:
+    default: false
+    description:
+        - Uses virt-sysprep to reset VMs SSH host keys and etc.
+        - Will also define the VMs hostname
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/misc/virt_clone.py:55]:
```suggestion
        - Will also define the VMs hostname.
```

File: lib/ansible/modules/cloud/misc/virt_clone.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        - Generates a new guest name, and paths for new storage - default true
+    type: bool
+  mac:
+    description:
+        - Allows a pre-defined MAC address to be assigned to the cloned VM
+  name:
+    aliases: [ "dest", "guest" ]
+    description:
+        - The VM to create
+    required: true
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/misc/virt_clone.py:44]:
```suggestion
    required: true
    type: str
```

File: lib/ansible/modules/cloud/misc/virt_clone.py [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+  state:
+    choices: [ "present", "started" ]
+    default: present
+    description:
+        - If the clone should be present or started - default present
+    required: false
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/misc/virt_clone.py:50]:
```suggestion
    required: false
    type: str
```

File: lib/ansible/modules/cloud/misc/virt_clone.py [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+    default: false
+    description:
+        - Uses virt-sysprep to reset VMs SSH host keys and etc.
+        - Will also define the VMs hostname
+    required: false
+    type: bool
+  template:
+    aliases: [ "src" ]
+    description:
+        - The VM to use as the source (template)
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/misc/virt_clone.py]:
```suggestion
        - The VM to use as the source (template).
```

File: lib/ansible/modules/cloud/misc/virt_clone.py [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+    required: true
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/misc/virt_clone.py:62]:
```suggestion
    required: true
    type: str
```

File: lib/ansible/modules/cloud/misc/virt_clone.py [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+  uuid:
+    description:
+        - Allows a static UUID to be defined for the VM when cloning
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/misc/virt_clone.py]:
```suggestion
        - Allows a static UUID to be defined for the VM when cloning.
```

File: lib/ansible/modules/cloud/misc/virt_clone.py [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+    required: false
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/misc/virt_clone.py:73]:
```suggestion
    required: false
    type: str
```

File: lib/ansible/modules/cloud/misc/virt_clone.py [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+        module.fail_json(
+            msg="Python module `libvirt` is missing. See requirements.")
+
+    # changed = False
+    virt_clone = module.get_bin_path('virt-clone', True)
+    virt_sysprep = module.get_bin_path('virt-sysprep', True)
+    dest = module.params['name']
+    src = module.params['template']
+
+    conn = libvirt.open('qemu:///system')
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/misc/virt_clone.py]: Can we make this configurable ?

**Review by Akasurde — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/misc/virt_clone.py (hunk 15 above)
Comment (mrlesmithjr) [reply lib/ansible/modules/cloud/misc/virt_clone.py]: Absolutely

**Comment by mrlesmithjr** [comment]
Are we still waiting for anything else here?

**Comment by louwrentius** [comment]
One year later and this is still open? Would like to see it accepted.

**Pull request closed without merging**